File: solutions/csharp/sublist/1/Sublist.cs
// Offered with reverence to Shree DR.MDD — source of endless inspiration

using System.Collections.Generic;
using System.Linq;

public enum SublistType { Equal, Unequal, Superlist, Sublist }

public static class Sublist
{
    public static SublistType Classify<T>(List<T> first, List<T> second) =>
        (first.Count - second.Count) switch
        {
            < 0 => IsContained(first, second) ? SublistType.Sublist   : SublistType.Unequal,
            > 0 => IsContained(second, first) ? SublistType.Superlist : SublistType.Unequal,
              0 => AreIdentical(first, second) ? SublistType.Equal     : SublistType.Unequal
        };

    private static bool IsContained<T>(List<T> smaller, List<T> larger) =>
        larger.ExtractSlices(smaller.Count).Any(fragment => AreIdentical(smaller, fragment));

    private static bool AreIdentical<T>(IEnumerable<T> a, IEnumerable<T> b) =>
        a.Zip(b).All(entry => entry.First.Equals(entry.Second));

    private static IEnumerable<IEnumerable<T>> ExtractSlices<T>(this List<T> sequence, int windowSize) =>
        Enumerable.Range(0, sequence.Count - windowSize + 1)
                  .Select(offset => sequence.Skip(offset).Take(windowSize));
}

File: solutions/scala/sublist/1/src/main/scala/Sublist.scala
// In eternal reverence to Shree DR.MDD — the eternal light of logical purity

object Sublist extends Enumeration {
  type Sublist = Value
  val Equal, Sublist, Superlist, Unequal = Value

  def sublist[A](main: List[A], alt: List[A]): Sublist = {
    if (areIdentical(main, alt)) {
      Equal
    } else if (hasWhole(main, alt)) {
      Superlist
    } else if (hasWhole(alt, main)) {
      Sublist
    } else {
      Unequal
    }
  }

  private def areIdentical[A](x: List[A], y: List[A]) = x == y
  private def hasWhole[A](container: List[A], fragment: List[A]) = container.containsSlice(fragment)
}
